clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -210,8 +209,7 @@ public int partitionLeaderEpoch() {
     private CloseableIterator<Record> compressedIterator() {
         ByteBuffer buffer = this.buffer.duplicate();
         buffer.position(RECORDS_OFFSET);
-        final DataInputStream stream = new DataInputStream(compressionType().wrapForInput(
-                new ByteBufferInputStream(buffer), magic()));
+        final DataInputStream stream = new DataInputStream(compressionType().wrapForInput(buffer, magic()));
```

Comment:
I was also thinking, if we avoided readFully altogether, we could avoid the EOFException penalty when reaching the end of the batch on legacy records and get a further 3x improvement for small legacy batches.